### PR TITLE
[CRITICAL] Implement <WorkspaceWorkflowsApprovalsExpensesFromPage /> for Member Selection

### PR DESCRIPTION
**Pull request by blazejkustra**
### Details

[Link to the doc](https://docs.google.com/document/d/12X-hsNo9BBRcG5NSS1JzyjWD1w-iM0uciQJSD1i5mZc/edit). 

Approval member page is going to be a page with a header and <SelectionList /> (multi select variant) with a save button.
image

Clicking on the next button will set the members with `setApprovalWorkflowMembers` action (more info in the [Onyx setup section](https://docs.google.com/document/d/12X-hsNo9BBRcG5NSS1JzyjWD1w-iM0uciQJSD1i5mZc/edit#heading=h.pdxemoaeemuw)). Users will be navigated to the Approver selection page (‘settings/workspaces/:policyID/workflows/approvals/approver’) when creating a new approval workflow or navigated to the  ‘Edit approval’ page (‘settings/workspaces/:policyID/workflows/approvals/:firstApproverEmail/edit’) based on `isBeingEdited` property in Onyx.

### Fixed Issues
$ https://github.com/Expensify/App/issues/45955
PROPOSAL: N/A

### Tests
- [x] Open Workflows page (it has to be turned on in `More features` settings page)
- [x] Locally edit `canUseWorkflowsAdvancedApproval` function to return true (or use `canUseWorkflowsAdvancedApproval` beta)
- [x] Click on the 'Add approval workflow' button, it should navigate to 'Expenses from' page
- [x] Verify that Expenses From page functions as other selection list (good example is Workspace invite screen)
  - "Find a member" input should filter users
  - You can select and deselect people in the list properly
- [x] Next button should navigate to the approver page

### Offline tests
N/A

### QA Steps
Same as tests

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

[android.webm](https://github.com/user-attachments/assets/e8ecca5c-9254-46ff-8b86-57c1033e78e0)

</details>

<details>
<summary>Android: mWeb Chrome</summary>

[android-web.webm](https://github.com/user-attachments/assets/44b821cf-dacd-48a7-afa9-a49530c3f5ef)

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/66d360e6-f756-4911-8b3b-cb5f2bbc6e73

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/user-attachments/assets/f5cae936-47ce-4bb2-ac90-9f5eea926fa9

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/50f72d95-a121-41c4-aa22-34f54b536f91

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/7495ebd2-cbff-412c-95f7-069e4b91fcf9

</details>

### CODE REVIEW

**Comment by Guccio163** [comment]
LGTM 🙌

**Review by Guccio163 — APPROVED** [review]
LGTM 🙌

File: src/languages/es.ts [hunk 1]
```diff
@@ -1291,6 +1291,7 @@ export default {
     },
     workflowsExpensesFromPage: {
         title: 'Gastos de',
+        header: 'Cuando los siguientes miembros presenten gastos:',
```
Comment (blazejkustra) [inline src/languages/es.ts:1294]: Translation already approved [here](https://docs.google.com/document/d/1Eh8MoQJ791xuxOEAlmAe1qHYPlh5UTUAHrGPrlV3088/edit?usp=sharing)

**Comment by blazejkustra** [comment]
@thesahindia I would appreciate a quick review on this one, as Approval Workflow project has a tight deadline (Aug 15)

**Comment by blazejkustra** [comment]
> After selecting the admin and clicking "Next", when we come back the "Admin" badge goes away.

On it!

> Shouldn't we dismiss the modal after the user selects an approver instead of navigating back to "Expenses from" page?

It will be covered in the next PR [here](https://github.com/Expensify/App/pull/46510), so I wouldn't block this PR because of it. (Add approval button is behind a beta so shouldn't be a problem).

**Comment by blazejkustra** [comment]
> We need to fix this error.
> 
> Screenshot 2024-07-30 at 7 24 02 PM

I don't see it, can you add reproduction steps? (make sure you have the newest changes)

**Comment by blazejkustra** [comment]
> After selecting the admin and clicking "Next", when we come back the "Admin" badge goes away.

@thesahindia Fixed!

**Review by thesahindia — APPROVED** [review]
👍🏼

**Review by tgolen — APPROVED** [review]
Looking good!

**Comment by kavimuru** [comment]
@tgolen @thesahindia How should we run this step?

> Locally edit canUseWorkflowsAdvancedApproval function to return true (or use canUseWorkflowsAdvancedApproval beta)

**Comment by tgolen** [comment]
I don't think this one is QA-able yet. Since it's under beta, the testing will come later. Sorry for the confusion! You can just mark this one off the deploy checklist.

File: src/pages/workspace/workflows/approvals/WorkspaceWorkflowsApprovalsExpensesFromPage.tsx [hunk 2]
```diff
@@ -1,28 +1,169 @@
...
+            Navigation.navigate(ROUTES.WORKSPACE_WORKFLOWS_APPROVALS_APPROVER.getRoute(route.params.policyID));
+        }
+    }, [approvalWorkflow?.approvers, approvalWorkflow?.isBeingEdited, route.params.policyID, selectedMembers]);
+
+    const goBack = useCallback(() => {
+        if (!approvalWorkflow?.isBeingEdited) {
+            Workflow.clearApprovalWorkflow();
+        }
+        Navigation.goBack();
+    }, [approvalWorkflow?.isBeingEdited]);
```
Comment (s77rt) [inline src/pages/workspace/workflows/approvals/WorkspaceWorkflowsApprovalsExpensesFromPage.tsx:145]: Clearing the approval workflow on going back caused the not found view to appear briefly while hiding the modal (https://github.com/Expensify/App/issues/50094). Also this `goBack` function is not called if we dismiss the modal on outside-click.

We went with a different approach that is to clear the workflow (from `WorkspaceWorkflowsPage`) before navigating to this page.

File: src/pages/workspace/workflows/approvals/WorkspaceWorkflowsApprovalsExpensesFromPage.tsx [hunk 3]
```diff
@@ -1,28 +1,169 @@
...
+        }
+
+        setSelectedMembers(
+            approvalWorkflow.members.map((member) => {
+                const policyMemberEmailsToAccountIDs = PolicyUtils.getMemberAccountIDsForWorkspace(policy?.employeeList);
+                const accountID = Number(policyMemberEmailsToAccountIDs[member.email] ?? '');
+                const isAdmin = policy?.employeeList?.[member.email].role === CONST.REPORT.ROLE.ADMIN;
+
+                return {
+                    text: member.displayName,
```
Comment (suneox) [inline src/pages/workspace/workflows/approvals/WorkspaceWorkflowsApprovalsExpensesFromPage.tsx:73]: Coming from #63939 checklist: The display name may include the SMS domain, so we should remove it to ensure consistency with InviteMemberListItem.

File: src/pages/workspace/workflows/approvals/WorkspaceWorkflowsApprovalsExpensesFromPage.tsx [hunk 4]
```diff
@@ -1,28 +1,169 @@
...
     // eslint-disable-next-line rulesdir/no-negated-variables
     const shouldShowNotFoundView = (isEmptyObject(policy) && !isLoadingReportData) || !PolicyUtils.isPolicyAdmin(policy) || PolicyUtils.isPendingDeletePolicy(policy);
 
+    useEffect(() => {
+        if (!approvalWorkflow?.members) {
+            return;
+        }
+
+        setSelectedMembers(
+            approvalWorkflow.members.map((member) => {
```
Comment (abzokhattab) [inline src/pages/workspace/workflows/approvals/WorkspaceWorkflowsApprovalsExpensesFromPage.tsx:67]: This introduced [#89270](https://github.com/Expensify/App/issues/89270). `getMemberAccountIDsForWorkspace(policy?.employeeList)` was called inside two separate `.map()` callbacks, so it ran once per member per render — O(N²) work. On workspaces with a few hundred members, opening "Add approval workflow" froze the app for roughly a minute. Fixed in [#89268](https://github.com/Expensify/App/pull/89268) by hoisting the call to the component body so it runs once per render.

**Pull request merged**